### DISCUSSION TITLE
The future of myanimetab

### DISCUSSION THREAD

**Discussion by Ari24-cb24** [aridevelopment-de/myanimetab#63] [Ideas]
# This discussion is locked. Please view the discussions created in the "Brainstorming" category

And here we start. The future of myanimetab has been something boggling me for a very long time now. I may have created a startpage for firefox but there is one key feature missing: anime. First of all, I didn't know in which direction to go. The structure of myanimetab is kind of messed up (considering I started this whole project in a sole html file) and there are lots of things I'm unhappy about. So in order to succeed with further progress, I have to

- Rewrite myanimetab
    - See #55, #49 and #30

Another important role plays the idea where to go with this project. As I said, anime is something I still want to implement. I considererd ideas such as

- Anime Search
- Weekly Anime (and notification subscriptions for them)
- Anime newsletter

and those ideas are great but that doesn't make us that unique as from the other competitors (tabliss, mytab, etc.). I gave it more and more thought and even thought about using myanimetab to teach japanese to weebs. And I think, I may try that out. My goal is to design the extension for weebs and thus teaching japanese is a great way of doing exactly that.

Long story short, I'll be using this discussion to notate the ideas I want to implement. The final decision in which direction myanimetab should go and what makes it unique. I won't really be able to work on new features before doing that (may be implementing something tiny) and I also consider a Rewrite of the whole project just to make it better.

**Comment by Ari24-cb24** [comment]
## Teaching japanese through myanimetab

Details tbd

**Comment by Ari24-cb24** [comment]
## Interactable waifus/husbandos

Similar to mahjong soul, you can have your own character with voice lines appearing here and there and telling some things. Maybe some kind of "Welcome back master" when the user visits the homepage (including a cooldown to not spam voice lines). Also, maybe some character interactions like specific games against a bot

See discussion #134

**Comment by Ari24-cb24** [comment]
## Anime Weekly List and Subscriptions

Details tbd

See discussion #135

**Comment by Ari24-cb24** [comment]
# General Rewrite

I don't like the current internal style of myanimetab as well as some external styles. This rewrite must
- [x] Use TypeScript
- [x] Use css module classes
- [ ] Include a complete set of tests
- [x] Making events more easier to access. 
- ~~Making widgets extend a class~~
- [x] Changing the style of the settings page
- [x] Adding an option for customizeable popup modals

**Comment by Ari24-cb24** [comment]
## Making events more easier to access

Atm the widget just stores the current setting value into the state and registers + unregisters an Event if the setting got updated and changes that setting + optionally doing some other updating. This should be completly removed and I propose an extendable class for those things. 

This class will have the following method: `registerInitial(state_key: str, settings_key: str, trigger_function: optional|function)`. This will retrieve the current value stored at `settings_key` and add an entry to the Component's state using `state_key` as the key. If the `settings_key` will change, the state automatically gets updated and thus rerenders the component with the new state. An optional function can be passed which will be triggered on that settings key update.

**Comment by Ari24-cb24** [comment]
## Making widgets extend a class

Idea deleted, see below

With the suggestion above, making Widgets extend a class. This class can have some optionables such as `Blurrable`, ... and will update the settings state accordingly.

**Comment by Ari24-cb24** [comment]
## Changing the style of the settings page

Changing style of the settings page. 

The first page will be the widget page but the widgets will be able to collapse/get an extra page if you click on them. So you're just able to see the widget with a toggle button. A `+` button will act as the "widget installer". With that you're able to add widgets as much as you want (2 clocks or more can be an option). This means that every widget will need their special own key. This will be pre-generated when extending from the widget class and will be used in `registerInitial` and (introducting another method) `updateSetting`. 

The second page will include the playlists and images. Here I want to follow a bit of the approach Wallpaperengine used meaning you can have multiple folders and have a root folder. Every image has their own checkbox and if you click on that, you will add that image to the current playlist. The current playlist can be toggled with a dropdown menu at the bottom of the page. To avoid a long page with images, multiple pages will be used. A future idea might be to add schedules when to switch playlists at specific times.

The third page might be either a "workshop collection" like the one from Wallpaper engine or a guide tab containing guides (like step-by-step learning-by-watching **interactable** tutorials). If the third page will be a workshop collection, optionally a fourth page will be introduced for the guides

**Comment by Ari24-cb24** [comment]
## Workshop Collections

This is something I've thought about since the beginning of myanimetab. Serving our own images to the user + letting users upload content to the workshop. I always avoided this since I worried about legal and licensing issues but I think if I get a bit more information, we should be able to do that. Image should normally be serverd as the urls instead of downloading the image and saving it on our server.  This will reduce traffic and disk space and might even avoid some legal issues. 

Letting users upload their own images might still be some kind of issue. Those images need to be verified beforehand (source and creator) and it's important that those images can be used for the application. I don't know if licensing can still get in the way if we just link the image from the original source but I should still worry about this.

See discussion #133

**Comment by Ari24-cb24** [comment]
## Adding an option for customizeable popup modals

Things like the fullscreen image, settings export/import and url add are being hardcoded with an event into the main class. This should be set to an event, carrying the component as the value

**Reply by Ari24-cb24** [reply]
This has been solved through mantine components

**Comment by Ari24-cb24** [comment]
## Using custom hooks

As React doesn't recommend extending components, we'll be using custom hooks instead. Two are called `useEvent` and `useSetting`. One updates the value when an event triggered and one when a setting changed. Because we're using hooks, every component has to be rewritten to a functional one.

**Comment by Ari24-cb24** [comment]
## Rendering Components

Atm the components itself keep track if they should render themselves. This should be abandoned and instead be implemented in the ComponentRenderer (currently the Main Component)

**Comment by Ari24-cb24** [comment]
## Widget mover

A long-time wish was to be able to just move widgets around when being in some kind of editing move. There would've been some snap-lines pre-configured and styling would've been done with percentages or pixels. The user could alternatively also edit the css (would be included in the settings) to set their own position.

See issue #119

**Comment by Ari24-cb24** [comment]
## Background Effects

This is a topic for another discussion but I'll mention it anyways here as this also belongs to the future of myanimetab. Similar to Wallpaper Engine, some custom background effect may be nice. I thought about moving things (leaves etc.) as well as bloom/lightness/vignette/... effects. This will get it's own editor.

See discussion #82 for more information

**Comment by Ari24-cb24** [comment]
I will now lock this conversation and create a new discussion thread for each of the elements that are still of interest
